pageserver/src/page_cache.rs

```diff
@@ -270,44 +272,29 @@ impl PageCache {
     //
     // Returns an 8k page image
     //
-    pub fn get_page_at_lsn(&self, tag: BufferTag, lsn: u64) -> anyhow::Result<Bytes> {
+    pub async fn get_page_at_lsn(&self, tag: BufferTag, lsn: u64) -> anyhow::Result<Bytes> {
         self.num_getpage_requests.fetch_add(1, Ordering::Relaxed);
 
         // Look up cache entry. If it's a page image, return that. If it's a WAL record,
         // ask the WAL redo service to reconstruct the page image from the WAL records.
         let minkey = CacheKey { tag, lsn: 0 };
         let maxkey = CacheKey { tag, lsn };
 
+        self.seqwait_lsn
+            .wait_for_timeout(lsn, TIMEOUT)
+            .await
+            .with_context(|| {
+                format!(
+                    "Timed out while waiting for WAL record at LSN {:X}/{:X} to arrive",
+                    lsn >> 32,
+                    lsn & 0xffff_ffff
+                )
+            })?;
+
         let entry_rc: Arc<CacheEntry>;
+
         {
-            let mut shared = self.shared.lock().unwrap();
-            let mut waited = false;
-
-            while lsn > shared.last_valid_lsn {
```

Comment:
Maybe keep an `assert!(lsn <= shared.last_valid_lsn)` here, to check that the value stored in the seqwait and `shared.last_valid_lsn` are in agreement.

Reply:
Note that @alubennikova added code here to skip the wait under certain circumstances. So this is only true if `walreceiver_works` is true.